### PR TITLE
Add DiffEqBase source support

### PR DESCRIPTION
**Pull request by davidanthoff**
Things to do:
- [x] Tests
- [x] ~~Docs~~ (nothing here has docs, so this can be merged without docs)
- [x] Don't hard code ``Float64`` (?)
- [x] Deal with solutions with multiple values
- [x] Use names from parameters for column headers if available

### CODE REVIEW

**Comment by davidanthoff** [comment]
It would be neat if one could collect into a ``TimeArray`` from [TimeSeries.jl](https://github.com/JuliaStats/TimeSeries.jl). For that the ``timestamp`` column that we generate here would have to be of some subtype of ``Base.Dates.TimeType``. Not sure whether ``DiffEqBase`` already supports this or how to do that.

File: src/integrations/differentialequations.jl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+function getiterator(sol::DiffEqBase.DESolution)
+    return DESolutionIterator(sol)
+end
+
+function Base.length(iter::DESolutionIterator)
+    return length(iter.sol)
+end
+
+function Base.eltype(iter::DESolutionIterator)
+    return @NT(timestamp::Float64, value::Float64)
```
Comment (ChrisRackauckas) [inline src/integrations/differentialequations.jl]: This won't always be Floats. At most you can expect `<:Number`.

**Comment by ChrisRackauckas** [comment]
>Deal with other solution types (?)

I think since you did it on the top abstract type, if this works then it will work for ODEs/RODEs/SDEs/DAEs/DDEs . For some of them, you might want to add extra fields, but that can come later. So if it can actually work at this level, that's great! They all have the `sol.t` and `sol.u` setup.

One thing to note is that in many cases someone may use a `ParameterizedFunction`, and there is a check `has_syms(sol.prob.f)` for whether an array `sol.prob.f.syms` exists. This array is an array of names that the user used when they created the system using one of the macros. Would it be possible to name the components using this array in this case? I was hoping to "name the columns" like in a DataFrame with these.

**Comment by ChrisRackauckas** [comment]
I plan on including this reexporting this library then with DifferentialEquations.jl once this is together, but I see this uses Requires.jl. Doesn't that library have issues with precompilation? I haven't touched it in awhile.

File: src/integrations/differentialequations.jl (hunk 1 above)
Comment (davidanthoff) [reply src/integrations/differentialequations.jl]: Alright, this should work with any type now.

**Comment by davidanthoff** [comment]
> One thing to note is that in many cases someone may use a ParameterizedFunction, and there is a check has_syms(sol.prob.f) for whether an array sol.prob.f.syms exists. This array is an array of names that the user used when they created the system using one of the macros. Would it be possible to name the components using this array in this case?

Would these names replace the column names ``timestamp`` and ``value`` in the code I'm using right now? I'm pretty sure that would work.

> I plan on including this reexporting this library then with DifferentialEquations.jl once this is together, but I see this uses Requires.jl. Doesn't that library have issues with precompilation? I haven't touched it in awhile.

Yes, right now it depends on Requires, but I *really* want to get rid of that at some point. The plan is to move all the integrations into their respective packages eventually, see #20. So ideally the code in this PR would eventually actually live in ``DiffEqBase``, and ``DiffEqBase`` would have a dependency on ``IterableTables``, which at that point would only be a few lines of code. But for now it is easier if I have everything in one package while I'm still tinkering with the overall design.

**Comment by ChrisRackauckas** [comment]
>Would these names replace the column names timestamp and value in the code I'm using right now? I'm pretty sure that would work.

Yes. `has_syms(sol.prob.f)` would be the bool for whether to explicitly name them, and `sol.prob.f.syms` is an array of names whose size matches `sol.u[i]` (the values at the `i`th timestep). Never used named tuples before so I'm not entirely sure how you'd do that.

>Yes, right now it depends on Requires, but I really want to get rid of that at some point. The plan is to move all the integrations into their respective packages eventually, see #20. So ideally the code in this PR would eventually actually live in DiffEqBase, and DiffEqBase would have a dependency on IterableTables, which at that point would only be a few lines of code. But for now it is easier if I have everything in one package while I'm still tinkering with the overall design.

Sounds good.

**Comment by davidanthoff** [comment]
So ``sol.u[i]`` would be an array if there are multiple values? Or would ``sol.u`` be a matrix in that case? To say it another way, ``sol.u`` is a vector of vectors if there are multiple values?

**Comment by ChrisRackauckas** [comment]
>To say it another way, sol.u is a vector of vectors if there are multiple values?

Yes.

**Comment by ChrisRackauckas** [comment]
Here's a good test:

```julia
using DiffEqBase, OrdinaryDiffEq, IterableTables, DataFrames

f_2dlinear = (t,u,du) -> du.=1.01u
prob = ODEProblem(f_2dlinear,rand(4,2),(0.0,1.0))
sol =solve(prob,Euler();dt=1//2^(4))

df = DataFrame(sol)

using ParameterizedFunctions

f = @ode_def LotkaVolterra begin
  dx = a*x - b*x*y
  dy = -c*y + d*x*y
end a=>1.5 b=>1 c=3 d=1

prob = ODEProblem(f,[1.0,1.0],(0.0,1.0))
sol =solve(prob,Tsit5())

df = DataFrame(sol)

has_syms(sol.prob.f) # true
sol.prob.f.syms # Symbol[:x,:y]
```

I'm not sure how to make a named tuple dynamically have names.

**Comment by davidanthoff** [comment]
> Here's a good test

Fantastic, that helped a lot!

> I'm not sure how to make a named tuple dynamically have names.

It is a delicate dance between writing a type stable iterator and constructing that in a dynamic way. I believe my latest push solves that. The simplicity is gone, of course, it now looks much more like the ``DataFrame`` code etc. But it works, I hope!

**Comment by ChrisRackauckas** [comment]
>But it works, I hope!

That seems to work great! Yes, not as simple haha, but man this is a really nice package. Does everything I was looking for. Good stuff.

>The plan is to move all the integrations into their respective packages eventually, see #20. So ideally the code in this PR would eventually actually live in DiffEqBase, and DiffEqBase would have a dependency on IterableTables, which at that point would only be a few lines of code. But for now it is easier if I have everything in one package while I'm still tinkering with the overall design.

I created a DiffEqIO.jl to hold all of the IO functionality of JuliaDiffEq. I think it's going to be a heavyweight dependency, so I'd like to isolate it a bit. I invited you to the team for it, and you have write capabilities so feel free to move this over at your own pace, or use it to experiment if you want.

**Comment by davidanthoff** [comment]
> I created a DiffEqIO.jl to hold all of the IO functionality of JuliaDiffEq. I think it's going to be a heavyweight dependency, so I'd like to isolate it a bit. I invited you to the team for it, and you have write capabilities so feel free to move this over at your own pace, or use it to experiment if you want.

Perfect, I'll migrate over once things are a bit more stable.

**Pull request merged**